public/app/plugins/datasource/elasticsearch/configuration/ElasticDetails.tsx

```diff
@@ -0,0 +1,202 @@
+                      {pattern.name}
+                    </option>
+                  ))}
+                </select>
+              }
+            />
+          </div>
+        </div>
+
+        <div className="gf-form max-width-25">
+          <FormField
+            labelWidth={10}
+            inputWidth={15}
+            label="Time field name"
+            value={value.jsonData.timeField || ''}
+            onChange={jsonDataChangeHandler('timeField', value, onChange)}
+            required
+          />
+        </div>
+
+        <div className="gf-form">
+          <span className="gf-form-select-wrapper">
+            <FormField
+              labelWidth={10}
+              label="Version"
+              inputEl={
+                <select
+                  className="gf-form-input gf-size-auto"
+                  value={value.jsonData.esVersion}
+                  onChange={jsonDataChangeHandler('esVersion', value, onChange)}
+                >
+                  {esVersions.map(version => (
+                    <option key={version.value} value={version.value}>
+                      {version.name}
+                    </option>
+                  ))}
+                </select>
+              }
+            />
+          </span>
+        </div>
+        {value.jsonData.esVersion >= 56 && (
+          <div className="gf-form max-width-30">
+            <FormField
+              aria-label={'Max concurrent Shard Requests input'}
+              labelWidth={15}
+              label="Max concurrent Shard Requests"
+              value={value.jsonData.maxConcurrentShardRequests || ''}
+              onChange={jsonDataChangeHandler('maxConcurrentShardRequests', value, onChange)}
+            />
+          </div>
+        )}
+        <div className="gf-form-inline">
+          <div className="gf-form">
+            <FormField
+              labelWidth={10}
+              label="Min time interval"
+              inputEl={
+                <Input
+                  className={'width-6'}
+                  value={value.jsonData.timeInterval || ''}
+                  onChange={jsonDataChangeHandler('timeInterval', value, onChange)}
+                  placeholder="10s"
+                  validationEvents={{
+                    [EventsWithValidation.onBlur]: [
+                      regexValidation(
+                        /^\d+(ms|[Mwdhmsy])$/,
+                        'Value is not valid, you can use number with time unit specifier: y, M, w, d, h, m, s'
+                      ),
+                    ],
+                  }}
+                />
+              }
+              tooltip={
+                <>
+                  A lower limit for the auto group by time interval. Recommended to be set to write frequency, for
+                  example <code>1m</code> if your data is written every minute.
+                </>
+              }
+            />
+          </div>
+        </div>
+      </div>
+    </>
+  );
+};
+
+const changeHandler = (
```

Comment:
This is interesting, I don't think i've seen having handlers defined outside of the scope of a component. It sort of makes sense, but is this primarily for making the component code more readable?

Reply:
Yeah readability mainly, I had it inside the function first as you would do it with `useCallback` but that was not needed for perf reasons so I tried putting it outside. But if you needed them memoized you would need to put them inside and use `useCallback` otherwise the memoization would have global cache which is not good for reuse. So I kinda like it but as it makes move to memoization harder (need to wrap __and__ move inside the function) maybe not something we should adopt in general.